ISSUE-3189 front end support for subaddressing

florentos17 opened the pull request:
cf
- [github issue tackled](https://github.com/linagora/tmail-flutter/issues/3189)
- [the jira ticket JAMES-3945 for subaddressing](https://issues.apache.org/jira/browse/JAMES-3945)
- [the github issue for right positioning on the backend side](https://github.com/linagora/james-project/issues/5289)
- [the documentation for the rights attribute of `Mailboxes`](https://github.com/linagora/james-project/blob/master/server/protocols/jmap-rfc-8621/doc/specs/spec/mail/rights.mdown)

florentos17 commented:
my plan in the following: `rights` is an attribute for mailboxes which indicates a list of rights for various identifiers, and in particular, the identifier `anyone`.

thus, subaddressing is considered allowed if the identifier `anyone` has the `p` right (`posting`). this `p` right does not exist just yet (as you can see in [the doc](https://github.com/linagora/james-project/blob/master/server/protocols/jmap-rfc-8621/doc/specs/spec/mail/rights.mdown)) but i will soon open a PR to add it. meanwhile, i try to set the `r` right.

clicking on the `allow/disallow subaddressing` menu option should create a JMAP request to set or remove the `p` right for the identifier `anyone`, but it does not work... i am able to effectively modify an other attribute (for example, `isSubscribed`), but it does not work for the attribute `rights`...

the interesting file is mailbox_api.dart lines 405-450.

<details>
  <summary>and here are the logs. as you can see, not much details...</summary>
  {using: [urn:ietf:params:jmap:mail, urn:ietf:params:jmap:core, urn:apache:james:params:jmap:mail:shares], methodCalls: [[Mailbox/set, {accountId: 548575abf99bc986f5f2318a4ab2ca22790513380dd5beb32daf004934e1635d, update:
{58fb3d50-8f84-11ef-87be-c1d0831d0b43: {isSubscribed: true}}}, c0]]}

*** Response ***
uri: https://jmap.lin-saas.dev/jmap
statusCode: 200
headers:
 access-control-allow-credentials: true
 access-control-allow-headers: Authorization
        DNT
        User-Agent
        X-Requested-With
        If-Modified-Since
        Cache-Control
        Content-Type
        Range
 access-control-allow-methods: GET
        POST
        PUT
        DELETE
 access-control-allow-origin: *
 access-control-expose-headers: **
 access-control-max-age: 1728000
 content-length: 331
 content-type: application/json
 date: Mon
         28 Oct 2024 14:26:17 GMT
 strict-transport-security: max-age=31536000; includeSubDomains
 vary: Origin

[TwakeMail] MailboxController::handleFailureViewState():Failure = SubaddressingMailboxFailure(null)

</details>

chibenwa commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +403,54 @@ class MailboxAPI with HandleSetErrorMixin {
+
+    updatedRights.update('anyone', (rights) {
+          if (action == MailboxSubaddressingAction.allow) {
+            rights?.add('r');
```
p ?

dab246 commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +403,54 @@ class MailboxAPI with HandleSetErrorMixin {
+        setMailboxInvocation.methodCallId,
+        SetMailboxResponse.deserialize);
+
+    return Future.sync(() async {
+      return setMailboxResponse?.updated?.containsKey(request.mailboxId.id) ?? false;
+    }).catchError((error) {
+      throw error;
+    });
```
we should not add try/catch at here. 
- should check condition and throw exception. 
- Should try/catch at DataSource

florentos17 replied:
you're right, there was already a try/catch block in MailboxDataSourceImpl

dab246 commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +403,54 @@ class MailboxAPI with HandleSetErrorMixin {
+        .toCapabilitiesSupportTeamMailboxes(session, accountId);
+
+    final response = await (requestBuilder
+      ..usings(capabilities))
```
indent

dab246 commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +403,54 @@ class MailboxAPI with HandleSetErrorMixin {
+  Map<String, List<String>?> _updateRightsForSubaddressing(MailboxSubaddressingAction action, Map<String, List<String>?>? currentRights) {
+    final updatedRights = Map<String, List<String>?>.from(currentRights ?? {});
+
+    updatedRights.update('anyone', (rights) {
+          if (action == MailboxSubaddressingAction.allow) {
+            rights?.add('r');
+          } else {
+            rights?.remove('r');
+          }
+          return rights;
+    },
+        ifAbsent: () => action == MailboxSubaddressingAction.allow ? ['r'] : null
+    );
```
Should separate the function to process it. And write unit test for it

dab246 replied:
And keywords (`anyone `, `r`) should be defined as `const`

dab246 commented on lib/features/mailbox/domain/state/subaddressing_mailbox_state.dart:
```diff
@@ -0,0 +1,34 @@
+import 'package:jmap_dart_client/jmap/core/state.dart' as jmap;
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subaddressing_action.dart';
+
+class LoadingSubaddressingMailbox extends UIState {}
```
Should extend from LoadingState

florentos17 replied:
ok ! to be honest i copied it from `lib/features/mailbox/domain/state/subscribe_mailbox_state.dart`, where there is `class LoadingSubscribeMailbox extends UIState {}`
what do you want me to do ?

dab246 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1336,6 +1353,60 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+  void _allowSubaddressingMailboxAction(MailboxId mailboxId, Map<String, List<String>?>? currentRights) {
+    final accountId = mailboxDashBoardController.accountId.value;
+    final session = mailboxDashBoardController.sessionCurrent;
+    if (session != null && accountId != null) {
+      final allowSubaddressingRequest = SubaddressingMailboxRequest(
+          mailboxId,
+          currentRights,
+          MailboxSubaddressingState.enabled,
+          MailboxSubaddressingAction.allow
+      );
+
+      consumeState(_subaddressingMailboxInteractor.execute(session, accountId, allowSubaddressingRequest));
+    }
```
Also should handle when `session` or `accountId` is null

dab246 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1336,6 +1353,60 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+  void _disallowSubaddressingMailboxAction(MailboxId mailboxId, Map<String, List<String>?>? currentRights) {
+    final accountId = mailboxDashBoardController.accountId.value;
+    final session = mailboxDashBoardController.sessionCurrent;
+    if (session != null && accountId != null) {
+      final disallowSubaddressingRequest = SubaddressingMailboxRequest(
+          mailboxId,
+          currentRights,
+          MailboxSubaddressingState.disabled,
+          MailboxSubaddressingAction.disallow
+      );
+
+      consumeState(_subaddressingMailboxInteractor.execute(session, accountId, disallowSubaddressingRequest));
+    }
```
idem

dab246 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1336,6 +1353,60 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+
+  void _handleSubaddressingMailboxFailure(SubaddressingMailboxFailure failure) {
+    if (currentOverlayContext != null && currentContext != null) {
+      var messageError = AppLocalizations.of(currentContext!).toastMessageSubaddressingFailure;
```
should use `final` replace for `var`

dab246 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1336,6 +1353,60 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+  }
+
+  void _handleSubaddressingMailboxSuccess(SubaddressingMailboxSuccess success) {
+    if (success.subaddressingAction == MailboxSubaddressingAction.allow) {
+      appToast.showToastSuccessMessage(
+          currentOverlayContext!,
+          AppLocalizations.of(currentContext!).toastMessageAllowSubaddressingSuccess);
+    } else {
+      appToast.showToastSuccessMessage(
+          currentOverlayContext!,
+          AppLocalizations.of(currentContext!).toastMessageDisallowSubaddressingSuccess);
+    }
```
```suggestion
appToast.showToastSuccessMessage(
  currentOverlayContext!,
  success.subaddressingAction == MailboxSubaddressingAction.allow
      ? AppLocalizations.of(currentContext!).toastMessageAllowSubaddressingSuccess
      : AppLocalizations.of(currentContext!).toastMessageDisallowSubaddressingSuccess,
);

```

dab246 commented on model/lib/extensions/presentation_mailbox_extension.dart:
```diff
@@ -49,6 +49,8 @@ extension PresentationMailboxExtension on PresentationMailbox {
 
   bool get isSubscribedMailbox => isSubscribed != null && isSubscribed?.value == true;
 
+  bool get isSubaddressingAllowed => rights != null && rights?['anyone']?.contains('p') == true;
```
And keywords (`anyone `, `p`) should be defined as `const`

florentos17 replied:
should i define those consts in a common location for both `presentation_mailbox_extension` and `mailbox_api` ?

dab246 replied:
IMO, we should do so

florentos17 commented on lib/features/mailbox/domain/model/mailbox_subaddressing_action.dart:
```diff
@@ -0,0 +1,5 @@
+enum MailboxSubaddressingAction {
+  allow,
+  disallow,
+  undo;
```
i implemented an undo button in the toast notification at first but i eventually chose not to and forgot to remove that line, thanks for spotting it

florentos17 commented on lib/features/mailbox/domain/model/subaddressing_request.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subaddressing_action.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subaddressing_state.dart';
+
+abstract class SubaddressingRequest with EquatableMixin {
```
it was not indeed

florentos17 commented on lib/features/mailbox/domain/state/subaddressing_mailbox_state.dart:
```diff
@@ -0,0 +1,34 @@
+
+class SubaddressingMailboxSuccess extends UIActionState {
+  final MailboxId mailboxId;
+  final MailboxSubaddressingAction subaddressingAction;
```
same response, i copied it from subscribe_mailbox_state.dart, but you are right i changed it.

florentos17 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1336,6 +1353,60 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
     }
   }
 
+  void _allowSubaddressingMailboxAction(MailboxId mailboxId, Map<String, List<String>?>? currentRights) {
+    final accountId = mailboxDashBoardController.accountId.value;
+    final session = mailboxDashBoardController.sessionCurrent;
+    if (session != null && accountId != null) {
+      final allowSubaddressingRequest = SubaddressingMailboxRequest(
+          mailboxId,
+          currentRights,
+          MailboxSubaddressingState.enabled,
+          MailboxSubaddressingAction.allow
+      );
+
+      consumeState(_subaddressingMailboxInteractor.execute(session, accountId, allowSubaddressingRequest));
+    }
+  }
+
+  void _disallowSubaddressingMailboxAction(MailboxId mailboxId, Map<String, List<String>?>? currentRights) {
+    final accountId = mailboxDashBoardController.accountId.value;
+    final session = mailboxDashBoardController.sessionCurrent;
+    if (session != null && accountId != null) {
+      final disallowSubaddressingRequest = SubaddressingMailboxRequest(
+          mailboxId,
+          currentRights,
+          MailboxSubaddressingState.disabled,
+          MailboxSubaddressingAction.disallow
+      );
+
+      consumeState(_subaddressingMailboxInteractor.execute(session, accountId, disallowSubaddressingRequest));
+    }
+  }
```
well, i copied it from `mailbox_controller.dart:_unsubscribeMailboxAction`, but i was wondering why both were necessary. should i keep only one (let's say `MailboxSubaddressingState` and try to completely remove the other ?

tddang-linagora replied:
These 2 methods perform an action, which need `SubaddressingMailboxRequest` as a parameter. Inside that request, consider if the action need to know current state of `MailboxSubaddressing, or/and the action need to be perform. If that action need both, why?

florentos17 replied:
the request carries information about the action to be performed, so the `subaddressingAction` parameter has to be kept.

`subaddressingState` was only used as a parameter in `SubaddressingSuccess` so that the toast notification could display the appropriate message. but `subaddressingAction` carries the same information as `subaddressingState` in the request, so i find it better to use `subaddressingAction` in `subaddressingSuccess` in order not to duplicate information.

doing so, `MailboxSubaddressingState` became completely unused, so i removed it.

florentos17 commented on lib/features/mailbox/domain/usecases/subaddressing_mailbox_interactor.dart:
```diff
@@ -0,0 +1,36 @@
+
+      final currentMailboxState = await _mailboxRepository.getMailboxState(session, accountId);
+
+      final result = await _mailboxRepository.subaddressingMailbox(session, accountId, request);
```
hmm more getting the status, to see whether there was a success or a failure. (i did the same as in `subscribe_mailbox_interactor`)

tddang-linagora replied:
Please rename the method `subaddressingMailbox` according to your intention

florentos17 commented:
the JMAP request generated by clicking the button still does not work. by looking into the james-project repo files [MailboxSet.scala](https://github.com/linagora/james-project/blob/master/server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/MailboxSet.scala) and [MailboxSetMethodContract.scala](https://github.com/linagora/james-project/blob/master/server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MailboxSetMethodContract.scala), i understood that the `rights` attribute is not meant to be modified directly, but rather using the `sharedWith/` keyword.

i managed to the the server response when sending such a request:
- with an incorrect right:
```
{"sessionState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","methodResponses":[["Mailbox/set",
{"accountId":"548575abf99bc986f5f2318a4ab2ca22790513380dd5beb32daf004934e1635d","oldState
":"a40c9441-9607-11ef-87be-c1d0831d0b43","newState":"a40c9441-9607-11ef-87be-c1d0831d0b43
","notUpdated":{"58fb3d50-8f84-11ef-87be-c1d0831d0b43":{"type":"invalidArguments","descri
ption":"Specified value do not match the expected JSON format:
List((/anyone(0),List(JsonValidationError(List(Unknown right
'y'),List()))))","properties":["sharedWith"]}}},"c0"]]}
```
- with a correct right:
```
{"sessionState":"2c9f1b12-b35a-43e6-9af2-0106fb53a943","methodResponses":[["Mailbox/set",{"accountId":"548575abf99bc986f5f2318a4ab2ca22790513380dd5beb32daf004934e1635d","oldState":"a40c9441-9607-11ef-87be-c1d0831d0b43","newState
":"a40c9441-9607-11ef-87be-c1d0831d0b43","notUpdated":{"58fb3d50-8f84-11ef-87be-c1d0831d0b43":{"type":"serverFail","description":null}}},"c0"]]}
```

i am stuck with this "serverFail" error... :/

chibenwa commented:
What is the corresponding JMAP request being done?

florentos17 commented:
> What is the corresponding JMAP request being done?

```
{using: [urn:ietf:params:jmap:mail, urn:ietf:params:jmap:core, urn:apache:james:params:jmap:mail:shares],
methodCalls: [[Mailbox/set, {accountId: 548575abf99bc986f5f2318a4ab2ca22790513380dd5beb32daf004934e1635d, update: {58fb3d50-8f84-11ef-87be-c1d0831d0b43: {sharedWith: {anyone: [r]}}}}, c0]]}
```

chibenwa commented:
Personal bet:
 - the issue comes from the fact that `anyone` isn't supported by the back
 - there is likely also something with non proper error handling going on

florentos17 commented:
i tried with other identifiers, including `user01`, `testuser01`, `bob`, `owner`, i always get the same `errorFail`. when i try with an empty identifier, i get "username should not be null or empty after being trimmed".

so indeed the error handling on the server side seems non exhaustive... is there any way i could access the logs of the server at https://jmap.lin-saas.dev to try to get more info ?

dab246 commented on core/lib/presentation/views/dialog/subaddressing_dialog_builder.dart:
```diff
@@ -0,0 +1,330 @@
+typedef OnCloseButtonAction = void Function();
+typedef OnCopyButtonAction = void Function();
+
+class SubaddressingDialogBuilder {
```
We already have such a dialog builder, why don't you reuse it?

florentos17 replied:
i want my popup to have two content "fields": some text ("You are about to..."), and below that, address display along with a `copy` button
![image](https://github.com/user-attachments/assets/3e322534-e227-48d4-b712-353d64bff43b)

i don't think the available dialog builders make something like this possible ?

dab246 replied:
You can still do it from our existing DialogBuilder abstract class. You just need to create a parameter for it and add a condition to render that widget. IMO, it quite easy, please take the time to research it.

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +597,18 @@ abstract class BaseMailboxController extends BaseController {
+        AppLocalizations.of(currentContext!).emailSubaddressCopiedToClipboard);
+  }
+
+  String getSubaddress(String userEmail, String folderName) {
+    int atIndex = userEmail.indexOf('@');
+    String localPart = userEmail.substring(0, atIndex);
+    String domain = userEmail.substring(atIndex + 1);
+    return "$localPart+$folderName@$domain";
```
if `atIndex=-1` it may throw an exception. You should write unit test for this function.

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +597,18 @@ abstract class BaseMailboxController extends BaseController {
+  void copySubaddressAction(BuildContext context, String subaddress) {
+    Clipboard.setData(ClipboardData(text: subaddress));
+    appToast.showToastSuccessMessage(
+        currentOverlayContext!,
+        AppLocalizations.of(currentContext!).emailSubaddressCopiedToClipboard);
```
You should use `context` function parameters

dab246 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -802,18 +820,22 @@ class ComposerController extends BaseController with DragDropFileMixin implement
         listToEmailAddress = List.from(recipients.value1.toSet());
         listCcEmailAddress = List.from(recipients.value2.toSet());
         listBccEmailAddress = List.from(recipients.value3.toSet());
+        listReplyToEmailAddress = List.empty(); //empty ? not sure about that
+        listReplyToEmailAddress = List.empty(); //empty ? not sure about that
```
remove duplicated

dab246 replied:
why is it empty?

florentos17 replied:
i left it empty at first because when composing a new mail, you would want the field `replyTo` to be initially empty. but digging further, there are test cases where it should not be empty: for example when editing a draft mail for which the `replyTo` field had been filled.
this is now fixed.

florentos17 commented:
@dab246 i think all of your comments were addressed, can i rebase and clean the commit history ?

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -442,6 +444,51 @@ abstract class BaseMailboxController extends BaseController {
+      String mailboxName,
+      String subaddress,
+      Map<String, List<String>?>? currentRights, {
+        required Function(MailboxId, Map<String, List<String>?>?, MailboxActions) onAllowAction
```
Should define name function for it to more clearly

dab246 replied:
Oh no, I mean: use `typedef` to define a function type, which makes the code more concise and readable. E.g:
```
typedef FunctionName = ReturnType Function(ParameterType1, ParameterType2, ...);
```

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +601,23 @@ abstract class BaseMailboxController extends BaseController {
+    appToast.showToastSuccessMessage(context, AppLocalizations.of(context).emailSubaddressCopiedToClipboard);
+  }
+
+  String getSubaddress(String userEmail, String folderName) {
+    if (folderName.isEmpty) {
+      throw const FormatException('folder name should not be empty');
+    }
+    int atIndex = userEmail.indexOf('@');
+    String localPart = userEmail.substring(0, atIndex);
+    String domain = userEmail.substring(atIndex + 1);
+    if (atIndex > 0 && localPart.isNotEmpty && domain.isNotEmpty) {
+      return "$localPart+$folderName@$domain";
+    } else {
+      throw const FormatException('Invalid email format');
+    }
+  }
```
To improve performance should use:

```suggestion
String getSubaddress(String userEmail, String folderName) {
  if (folderName.isEmpty) {
    throw FormatException('Folder name should not be empty');
  }

  final atIndex = userEmail.indexOf('@');
  if (atIndex <= 0 || atIndex == userEmail.length - 1) {
    throw FormatException('Invalid email format');
  }

  return '${userEmail.substring(0, atIndex)}+$folderName@${userEmail.substring(atIndex + 1)}';
}
```

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +601,23 @@ abstract class BaseMailboxController extends BaseController {
+
+  String getSubaddress(String userEmail, String folderName) {
+    if (folderName.isEmpty) {
+      throw const FormatException('folder name should not be empty');
```
Should define a separate exception for it. It could be: `FolderEmptyNameException`

dab246 commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +404,52 @@ class MailboxAPI with HandleSetErrorMixin {
+    if (setMailboxResponse?.updated?.containsKey(request.mailboxId.id) ?? false) {
+      return true;
+    } else {
+      throw Exception("Failed to update mailbox rights.");
```
Should define a separate exception for it.

dab246 commented on test/features/mailbox/data/update_rights_for_subaddressing_test.dart:
```diff
@@ -0,0 +1,55 @@
+    const postingRight = 'p';
+    final mailboxAPI = MailboxAPI(HttpClient(Dio()), Uuid());
+
+
```
remove line

dab246 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1336,6 +1367,46 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+
+      consumeState(_subaddressingInteractor.execute(session, accountId, allowSubaddressingRequest));
+    } else {
+      _handleSubaddressingFailure(SubaddressingFailure("session and accountId should not be null"));
```
Should define a separate exception for it

dab246 commented on lib/features/mailbox/presentation/widgets/copy_subaddress_widget.dart:
```diff
@@ -0,0 +1,53 @@
+class CopySubaddressWidget extends StatelessWidget {
+
+  final BuildContext context;
+  final ImagePaths imagePath;
+  final String subaddress;
+
+  final OnCopyButtonAction onCopyButtonAction;
+
+  const CopySubaddressWidget({
+    super.key,
+    required this.context,
+    required this.imagePath,
+    required this.subaddress,
+    required this.onCopyButtonAction
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    final imagePaths = Get.find<ImagePaths>();
```
remove it if not use

dab246 commented:
> @dab246 i think all of your comments were addressed, can i rebase and clean the commit history ?

Yes, I have closed the previous comments and reviewed them again. Please fix the remaining comments. Thanks.

florentos17 commented:
> Yes, I have closed the previous comments and reviewed them again. Please fix the remaining comments. Thanks.

all done !

florentos17 commented:
i rebased and cleaned the commit history, unfortunately i seem to have some check failures ("Error: The process '/opt/hostedtoolcache/flutter/bin/cache/dart-sdk/bin/dart' failed with exit code 2") but i am not sure what this means or where this comes from ?

tddang-linagora commented:
>but i am not sure what this means or where this comes from ?

![Screenshot 2024-11-13 at 10 26 33](https://github.com/user-attachments/assets/bf474e2b-0154-4d02-92a5-2b310801c8ab)

Look at the warnings and infos, if its cause is from your code, please resolve it.

dab246 commented:
- [ ] Please fix build CI failed

<img width="1427" alt="Screenshot 2024-11-14 at 09 30 00" src="https://github.com/user-attachments/assets/cc9ae54e-5b68-4e4e-a4ee-d376d92c6822">

florentos17 commented:
> * [ ]  Please fix build CI failed

- i am quite confused about the first three errors (`The getter 'pendingFileInfo' isn't defined for the type 'MockEmailReceiveManager'`): the getters are actually defined, and called successfully. i do not have that error locally.
- the last error (in `identity_creator_controller`) i do have locally, but i have not edited that file, so i don't think i am responsible for it ?

dab246 commented:
> > * [ ]   Please fix build CI failed
> 
> * i am quite confused about the first three errors (`The getter 'pendingFileInfo' isn't defined for the type 'MockEmailReceiveManager'`): the getters are actually defined, and called successfully. i do not have that error locally.
> * the last error (in `identity_creator_controller`) i do have locally, but i have not edited that file, so i don't think i am responsible for it ?

IMO, you should rebase on master then check again.

florentos17 commented:
we want to use URL encode for the local part details of email addresses (cf [this ticket](https://github.com/linagora/james-project/issues/5296))
i feel like what i did should work, however when i test it, sometimes it works, sometimes it does not, and i cant understand exactly why just yet... i am still investigating it.

florentos17 commented on core/lib/utils/mail/mail_address.dart:
```diff
@@ -137,6 +145,54 @@ class MailAddress with EquatableMixin {
+    return MailAddress(localPart: getLocalPartWithoutDetails(), domain: domain);
+  }
+
+  String strictUrlEncode(String input) {
+    final StringBuffer encoded = StringBuffer();
+
+    for (int codeUnit in utf8.encode(input)) {
+      if ((codeUnit >= 0x30 && codeUnit <= 0x39) || // 0-9
+          (codeUnit >= 0x41 && codeUnit <= 0x5A) || // A-Z
+          (codeUnit >= 0x61 && codeUnit <= 0x7A)) { // a-z
+        encoded.writeCharCode(codeUnit);
+      } else {
+        encoded.write('%${codeUnit.toRadixString(16).padLeft(2, '0').toUpperCase()}');
+      }
+    }
+    return encoded.toString();
+  }
```
i tried it with `Uri.encodeComponent()` at first, but the characters `-_.!~*'()` are not encoded, that's why i rewrote it entirely.

chibenwa replied:
Where is the upstream repo for this method?
Did we open a bug report?
And maybe even an upstream PR too?
And we should put the link of all this upstream work here.

florentos17 replied:
this method i wrote myself

you can find [Uri.encodeComponent() here](https://github.com/dart-lang/sdk/blob/808fa4ca8b5fbc3d79ade02f607b8533c9714029/sdk/lib/core/uri.dart), where they specify that `-_.!~*'()` are left as-is because it's meant to use in the context of RFC 2396.
so there is not really a bug to report right ?

chibenwa replied:
>  so there is not really a bug to report right ?

Maybe not then but at least a code comment.

chibenwa commented on core/lib/utils/mail/mail_address.dart:
```diff
@@ -166,6 +166,8 @@ class MailAddress with EquatableMixin {
   }
 
   String strictUrlEncode(String input) {
+    // cannot use Uri.encodeComponent because it is meant to be compliant with RFC2396
```
Link where the code is copied from please

chibenwa commented on core/lib/utils/mail/mail_address.dart:
```diff
@@ -189,7 +222,7 @@ class MailAddress with EquatableMixin {
       return MailAddress.fromParts(
           localPartWithoutDetails: getLocalPartWithoutDetails(),
-          localPartDetails: strictUrlEncode(localPartDetails),
+          localPartDetails: customUriEncode(localPartDetails),
```
Are we sure this is 100% compliant with what is implemented on James server side?

hoangdat commented:
hi @florentos17  we have `reply-to` in composer, but can we also support AutoComplete with subAddressing. IMO it will be easier for user to input on it.

hoangdat commented:
- [ ] Mailbox set still failed with `allow subadressing` in `jmap.lin-saas.dev` platform
![image](https://github.com/user-attachments/assets/4723efca-e6a7-4b41-a02a-068b4a41fe05)

hoangdat commented:
- [ ] if sub addressing is not allowed, so I think we should prevent user to copy it? WDYT @florentos17

florentos17 commented:
> * [ ]  if sub addressing is not allowed, so I think we should prevent user to copy it? WDYT @florentos17

are you talking about the "copy address" in the options menu ? if so, yes i agree it makes sense

florentos17 commented:
here is a pretty good idea of what the final version could look like ! don't hesitate if you have comments !

basic demo:
https://github.com/user-attachments/assets/81966af2-0a79-477f-aae1-bc648681cfae

advanced demo:
https://github.com/user-attachments/assets/02be6f88-73b2-4e7a-b79d-c29af7e51e79

florentos17 commented:
[subaddressing-demo.webm](https://github.com/user-attachments/assets/312da4bb-82a0-4771-860b-6c388422feff)

florentos17 commented:
[subaddressing-demo-advanced.webm](https://github.com/user-attachments/assets/75e5b676-8632-44a9-974e-320c068a662d)

chibenwa commented:
Good quality work done here!

chibenwa commented:
@hoangdat ok to you?

hoangdat commented:
Hi, srr for late response, but today it is quite overload for me in both tmail and tchat. 

For this, I still not deep in the implementation, and still have a lot of concerned relate to UXUI.

hoangdat commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +405,52 @@ class MailboxAPI with HandleSetErrorMixin {
   }
 
+  List<String> updateRightsForSubaddressing(MailboxSubaddressingAction action, List<String>? currentRights) {
+    const String postingRight = 'p';
```
please define `const` in global

florentos17 replied:
both `postingRight` and `anyoneIdentifier` are needed by two files:
- `lib/features/mailbox/data/network/mailbox_api.dart`
- `model/lib/extensions/presentation_mailbox_extension.dart`

should i create some sort of `constants.dart` file to define them ?

hoangdat commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +405,52 @@ class MailboxAPI with HandleSetErrorMixin {
+  }
+
+  Future<bool> handleSubaddressingRequest(Session session, AccountId accountId, SubaddressingRequest request) async {
+    const String anyoneIdentifier = 'anyone';
```
please define const in global

florentos17 replied:
done !

hoangdat commented on lib/features/mailbox/domain/model/subaddressing_request.dart:
```diff
@@ -0,0 +1,23 @@
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subaddressing_action.dart';
+
+class SubaddressingRequest with EquatableMixin {
```
```suggestion
class MailboxRightRequest with EquatableMixin {
```

florentos17 replied:
done !

hoangdat commented on lib/features/mailbox/presentation/mixin/mailbox_widget_mixin.dart:
```diff
@@ -60,6 +60,12 @@ mixin MailboxWidgetMixin {
         MailboxActions.markAsRead,
       MailboxActions.move,
       MailboxActions.rename,
+      if (mailbox.isSubaddressingAllowed)
+        MailboxActions.disallowSubaddressing
+      else
+        MailboxActions.allowSubaddressing,
+      if (mailbox.isSubaddressingAllowed)
+        MailboxActions.copySubaddress,
```
we should check if capability `urn:apache:james:params:jmap:mail:shares` available or not. Please refer to the usage of `requireCapability` or `capability_validator.dart` for more information.

florentos17 replied:
done !

hoangdat commented:
- [ ] a lot of padding in this dialog make unnecessary space
![image](https://github.com/user-attachments/assets/b6387408-6f7d-41ae-9740-7fb2d25cd245)

hoangdat commented:
- [ ] when someone send mail to subaddressing, it is not route to specific folder

![image](https://github.com/user-attachments/assets/81543177-2829-4bdd-9d62-0d453201ea63)

dab246 commented on core/lib/utils/mail/mail_address.dart:
```diff
@@ -416,6 +510,30 @@ class MailAddress with EquatableMixin {
+    }
+
+    String localPartDetails = localPartDetailsSB.toString();
+    // should validate localPartDetails
```
IMO, it doesn't make sense

florentos17 replied:
i am meaning to come back to this comment: currently any string will be accepted as a localPartDetails, but the string should follow the same rules as the name of a folder, for example `user+folder%name@domain.com` cannot be valid since no folder can be named `folder%name`

dab246 commented on test/features/composer/presentation/composer_controller_test.dart:
```diff
@@ -285,6 +285,7 @@ void main() {
       final toRecipient = EmailAddress('to', 'to@linagora.com');
       final ccRecipient = EmailAddress('cc', 'cc@linagora.com');
       final bccRecipient = EmailAddress('bcc', 'bcc@linagora.com');
+      final replyToRecipient = EmailAddress('replyTo', 'bcc@linagora.com');
```
```suggestion
      final replyToRecipient = EmailAddress('replyTo', 'replyTo@linagora.com');
```

dab246 commented:
- [ ] On web responsive mobile, the `Reply To` button doesn't hide after click on. 



https://github.com/user-attachments/assets/1f2e729b-246b-4efd-add0-6076276e7913

dab246 commented:
- [ ] The `Allow subaddressing` icon is blurred 


<img width="540" alt="Screenshot 2024-11-21 at 14 37 34" src="https://github.com/user-attachments/assets/44773b5f-f1d3-4f80-968e-98d434771e2b">

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +611,22 @@ abstract class BaseMailboxController extends BaseController {
+
+    final atIndex = userEmail.indexOf('@');
+    if (atIndex <= 0 || atIndex == userEmail.length - 1) {
+      throw const FormatException('Invalid email format');
```
Should create class for it same with `EmptyFolderNameException `

florentos17 replied:
done !

dab246 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1338,6 +1369,47 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+      consumeState(_subaddressingInteractor.execute(session, accountId, allowSubaddressingRequest));
+    } else {
+      _handleSubaddressingFailure(SubaddressingFailure.withException(
+          SubaddressingException("Session and accountId should not be null")));
```
Should create specific class for this exception

florentos17 replied:
done !

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -442,6 +454,51 @@ abstract class BaseMailboxController extends BaseController {
     }
   }
 
+  void openConfirmationDialogSubaddressingAction(
```
Should be moved to `mailbox controller`

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +611,22 @@ abstract class BaseMailboxController extends BaseController {
     return mailboxNode;
   }
+
+  void copySubaddressAction(BuildContext context, String subaddress) {
```
idem

dab246 commented on lib/features/base/base_mailbox_controller.dart:
```diff
@@ -554,4 +611,22 @@ abstract class BaseMailboxController extends BaseController {
+    appToast.showToastSuccessMessage(context, AppLocalizations.of(context).emailSubaddressCopiedToClipboard);
+  }
+
+  String getSubaddress(String userEmail, String folderName) {
```
idem

florentos17 commented on assets/images/ic_subaddressing_allow.svg:
i asked Bob PODVALNYI for them (bpodvalnyi@linagora.com). should i use M3 icons instead as suggested in your comment above ?

hoangdat replied:
Maybe your export/import make the icon break. If you can find the same in M3, please use. If not, please make sure the format of icon still ok.

florentos17 replied:
the copy icon is easily replaceable, and for the other two i found those:

![Screenshot from 2024-11-21 11-43-37](https://github.com/user-attachments/assets/caeaeb45-9e6e-4390-b192-724bf19cb912)
![Screenshot from 2024-11-21 11-43-44](https://github.com/user-attachments/assets/5346001b-37da-498c-8cb9-2aec5b5d586e)

i feel like they would not integrate as well with the other icons of the menu, but it's not a big difference. what do you think ?

florentos17 replied:
the icons were broken from the start, but bob redid them for me, now its fixed !

dab246 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -1567,9 +1614,20 @@ class ComposerController extends BaseController
       .contains(inputEmail);
   }
 
+  String? getDisplayName(MailAddress mailAddress) {
+    String? localPartDetails = mailAddress.getLocalPartDetails();
+    if(localPartDetails == null) {
+      return null;
+    } else {
+      return '${mailAddress.getLocalPartWithoutDetails()} [${mailAddress.getLocalPartDetails()}]';
+    }
+  }
```
```suggestion
extension MailAddressExtension on MailAddress {
  String? get getDisplayName {
    String? localPartDetails = getLocalPartDetails();
    if(localPartDetails == null) {
      return null;
    } else {
      return '${getLocalPartWithoutDetails()} [${getLocalPartDetails()}]';
    }
  }
}
```

florentos17 replied:
done !

dab246 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -1583,8 +1641,10 @@ class ComposerController extends BaseController
   void _autoCreateCcEmailTag(String inputEmail) {
-    if (!_isDuplicatedRecipient(inputEmail, listCcEmailAddress)) {
-      final emailAddress = EmailAddress(null, inputEmail);
+    MailAddress inputAddress = MailAddress.validateAddress(inputEmail);
+    String encodedInputEmail = inputAddress.asStringWithEncodedLocalPartDetails();
+    if (!_isDuplicatedRecipient(encodedInputEmail, listCcEmailAddress)) {
+      final emailAddress = EmailAddress(getDisplayName(inputAddress), encodedInputEmail);
```
idem

florentos17 replied:
done !

dab246 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -1597,8 +1657,10 @@ class ComposerController extends BaseController
   void _autoCreateBccEmailTag(String inputEmail) {
-    if (!_isDuplicatedRecipient(inputEmail, listBccEmailAddress)) {
-      final emailAddress = EmailAddress(null, inputEmail);
+    MailAddress inputAddress = MailAddress.validateAddress(inputEmail);
+    String encodedInputEmail = inputAddress.asStringWithEncodedLocalPartDetails();
+    if (!_isDuplicatedRecipient(encodedInputEmail, listBccEmailAddress)) {
+      final emailAddress = EmailAddress(getDisplayName(inputAddress), encodedInputEmail);
```
idem

florentos17 replied:
done !

dab246 commented:
- To display email display name with folder name as below picture:

<img width="369" alt="Screenshot 2024-11-21 at 16 10 12" src="https://github.com/user-attachments/assets/cb13e269-4d88-4a60-bec9-4441adea860b">

You need change some class: `SuggestionEmailAddress`, `DraggableEmailAddress`, 

or functions:

![Screenshot 2024-11-21 at 16 08 26](https://github.com/user-attachments/assets/ad13014b-595d-4dac-b743-ddbc3935ebed)

florentos17 commented on lib/features/composer/presentation/styles/composer_style.dart:
```diff
@@ -155,10 +155,6 @@ class ComposerStyle {
-      return constraints.maxHeight > 0 ? constraints.maxHeight * 0.4 : 150.0;
-    }
+    return constraints.maxHeight > 0 ? constraints.maxHeight * 0.4 : 150.0;
   }
```
[Screencast from 21-11-2024 11:26:17.webm](https://github.com/user-attachments/assets/4897a11d-29f6-415c-9892-22540452e9a5)

when clicking all options in the composer (`from`, `to`, `cc`, `bcc` and the new one `replyTo`), some scrolling was needed (cf video above), and that change made it possible to view all of them at once (cf screenshot below). is that ok ?

![image](https://github.com/user-attachments/assets/70b2e956-6063-40cf-bdcf-40488e66764a)

florentos17 commented on lib/features/mailbox/presentation/mixin/mailbox_widget_mixin.dart:
```diff
@@ -248,6 +274,23 @@ mixin MailboxWidgetMixin {
+      return false;
+    }
+
+    final subaddressingSupported = ((session
+        .props[0] as Map<CapabilityIdentifier, CapabilityProperties>?)
+        ?[capabilityDeletedMessagesVault]
+        ?.props[0] as Map<String, dynamic>?)
+        ?['subaddressingSupported'];
+
+    return subaddressingSupported ?? true;
```
this check is related to [this new ticket](https://github.com/linagora/tmail-backend/issues/1324): we only want the subaddressing options to be available if the server explicitly advertizes that it supports it.

since it does not for now, the check will always fail, so i leave the default value to `true` for testing purposes. this will have to be changed to `false` before merging

florentos17 commented:
> * [ ]  when someone send mail to subaddressing, it is not route to specific folder
> 
> 
> ![image](https://private-user-images.githubusercontent.com/6462404/388389943-81543177-2829-4bdd-9d62-0d453201ea63.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyMDYyMDQsIm5iZiI6MTczMjIwNTkwNCwicGF0aCI6Ii82NDYyNDA0LzM4ODM4OTk0My04MTU0MzE3Ny0yODI5LTRiZGQtOWQ2Mi0wZDQ1MzIwMWVhNjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjFUMTYxODI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTczNmEzMzkyYTQzNzM5ZTRiYTI2MTVmYzU4ZTczYWRmNGFkODkzNDkzMzJkOTdiOWM0ZGI0Y2E0NGQ3YWFiMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.kJ_grdyeF_ShZkHySt6h9EMuhJHDwuRcGn4HgUKvgKQ)

can you tell me more about the context of this test:
- do you have a folder named `Apple` ?
- did you enable subadressing for this folder before sending the mail ?
- was the backend server you were connected to supporting subaddressing ?

hoangdat commented:
> * do you have a folder named `Apple` ?

Yes

> * did you enable subadressing for this folder before sending the mail ?

Yes
 
> * was the backend server you were connected to supporting subaddressing ?

tmail.linagora.com

hoangdat commented on lib/features/mailbox/data/network/mailbox_api.dart:
```diff
@@ -401,6 +405,52 @@ class MailboxAPI with HandleSetErrorMixin {
     return listMailboxIdSubscribe ?? [];
   }
 
+  List<String> updateRightsForSubaddressing(MailboxSubaddressingAction action, List<String>? currentRights) {
```
yeah, no way to test private method, but we should test the public method which invoke the private one.

hoangdat commented on core/lib/presentation/views/dialog/confirmation_dialog_builder.dart:
```diff
@@ -233,6 +238,11 @@ class ConfirmDialogBuilder {
               ),
+            if (_additionalWidgetContent != null)
+              Padding(
+                padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
```
_textContent and _additionalWidgetContent also have padding, make a lot of space between it. No need `top = 8` here.

florentos17 replied:
done !

hoangdat commented on model/lib/extensions/presentation_email_extension.dart:
```diff
@@ -120,27 +120,27 @@ extension PresentationEmailExtension on PresentationEmail {
   }
 
-  Tuple3<List<EmailAddress>, List<EmailAddress>, List<EmailAddress>> generateRecipientsEmailAddressForComposer({
+  Tuple4<List<EmailAddress>, List<EmailAddress>, List<EmailAddress>, List<EmailAddress>> generateRecipientsEmailAddressForComposer({
```
please write test for this.

florentos17 replied:
done !

dab246 commented:
@florentos17 Please fix all comments ASAP. For any comments that have been resolved, kindly leave a reply so we can `resolve` them. Thanks!

tddang-linagora commented:
- [ ] Update the PR description with a working demo, preferably with responsive/multiple resolutions.

florentos17 commented on test/features/mailbox/repository/mailbox_respository_test.dart:
```diff
@@ -229,4 +236,47 @@ void main() {
+        when(mailboxDataSource.handleMailboxRightRequest(sessionFixture, accountIdFixture, request))
+          .thenAnswer((_) async => true);
+
+        when(mailboxDataSource.getChanges(sessionFixture, accountIdFixture, StateFixtures.currentMailboxState))
+          .thenAnswer((_) async => MailboxChangeResponse(
+              // dont really know what to put in there
+              updated: [mailboxFixture],
+              updatedProperties: Properties({'rights'})
+        ));
+
+        final response = await mailboxDataSource.getChanges(
+          sessionFixture,
+          accountIdFixture,
+          StateFixtures.currentMailboxState,
+        );
+
+        expect(response.props[6], Properties({'rights'}));
```
here i want to write a test for `updateRightsForSubaddressing`, which will be turned private, so the test is done via the caller method `handleMailboxRightRequest`.

i tried to do something based on what i could find, but i dont really understand how these kind of tests work (i am basically expecting on line 279 what i am inputing on line 269, so of course it works but it does not test anything...)

florentos17 replied:
dat dang suggested to add this in order to be able to test `updateRightsForSubaddressing` directly, it works fine if you're ok with it:

```
@visibleForTesting
List<String> updateRightsForSubaddressing(MailboxSubaddressingAction action, List<String>? currentRights)
    => _updateRightsForSubaddressing(action, currentRights);
```

florentos17 commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1061,13 +1072,33 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+      case MailboxActions.copySubaddress:
+        copySubaddressAction(
+            context,
+            getSubaddress(mailboxDashBoardController.userEmail, findNodePathWithSeparator(mailbox.id, '.')!));
```
done !

florentos17 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -1567,9 +1614,20 @@ class ComposerController extends BaseController
+    MailAddress inputAddress = MailAddress.validateAddress(inputEmail);
+    String encodedInputEmail = inputAddress.asStringWithEncodedLocalPartDetails();
+    if (!_isDuplicatedRecipient(encodedInputEmail, listToEmailAddress)) {
+      final emailAddress = EmailAddress(getDisplayName(inputAddress), encodedInputEmail);
```
done !

florentos17 commented on assets/images/ic_subaddressing_disallow.svg:
icon fixed

hoangdat commented on lib/features/base/upgradeable/upgrade_hive_database_steps_v13.dart:
```diff
@@ -0,0 +1,17 @@
+  @override
+  Future<void> onUpgrade(int oldVersion, int newVersion) async {
+    if (oldVersion > 0 && oldVersion < newVersion && newVersion == 13) {
+      await _cachingManager.clearEmailCacheAndAllStateCache();
```
please also delete cache for mailbox.

hoangdat commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1061,13 +1076,41 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
         break;
+      case MailboxActions.copySubaddress:
+        try{
+          final subaddress = getSubaddress(mailboxDashBoardController.userEmail, findNodePathWithSeparator(mailbox.id, '.')!);
```
When I copied this subaddress: `dphamhoang+subInbox hfjdhdh hfjdhdh  dhhdhdhd hfhfhf.ti disable .ddedeutuuiiiosoodsodo oodsoidoisooo dosdisodiosoid dsoidosod @linagora.com`, composer said it error email address.

<img width="881" alt="image" src="https://github.com/user-attachments/assets/9671fa1c-c875-45a6-9129-dd2dc3048c82">

chibenwa replied:
Yes space shall be encoded.

florentos17 replied:
it was correctly encoded but not recognized by the composer (`%` was a forbidden character), this is fixed now

hoangdat replied:
should have test cases for it.

florentos17 replied:
already done in `mail_address_test.dart`

hoangdat commented on lib/features/mailbox/presentation/mailbox_controller.dart:
```diff
@@ -1338,6 +1381,46 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
+          : AppLocalizations.of(currentContext!).toastMessageDisallowSubaddressingSuccess,
+    );
+
+    _refreshMailboxChanges(
+        currentMailboxState: success.currentMailboxState,
+        properties: MailboxConstants.propertiesDefault
+    );
```
IMO, it is an unnecessary call. Duplicated refresh request.

<img width="497" alt="Screenshot 2024-11-28 at 11 47 47" src="https://github.com/user-attachments/assets/718b5967-5c37-4693-9066-da0ef0d39803">

hoangdat commented:
### Target for pairing
- [ ] how to test sub-addressing (tmail.linagora can not work when sending email to subaddress email)
- [ ] generating MailAddress for sub-address (encode, decode) + testing all cases (https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1861499772)
- [ ] mailbox cache upgrading (https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1861491709)
- [ ] duplicated call for refresh mailbox after allow sub-addressing (https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1861503085)

@tddang-linagora @florentos17

tddang-linagora reviewed (dismissed):
- [ ] Resolve CI

florentos17 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -105,6 +105,21 @@ import 'package:tmail_ui_user/main/localizations/app_localizations.dart';
 import 'package:tmail_ui_user/main/routes/route_navigation.dart';
 import 'package:tmail_ui_user/main/universal_import/html_stub.dart' as html;
 
+extension MailAddressExtension on MailAddress {
+  String? get getDisplayName {
+    String? localPartDetails = getLocalPartDetails();
+    if(localPartDetails == null) {
+      return null;
+    } else {
+      return '${getLocalPartWithoutDetails()} [${getLocalPartDetails()}]';
+    }
+  }
+
+  EmailAddress asEmailAddress() {
+    return EmailAddress(getDisplayName, asStringWithEncodedLocalPartDetails());
+  }
+}
+
```
done !

hoangdat commented:
> i think all the comments were addressed. can i clean the commit history and rebase ?

Yes, move on

hoangdat commented on test/features/mailbox/repository/mailbox_respository_test.dart:
```diff
@@ -229,4 +236,47 @@ void main() {
+        when(mailboxDataSource.handleMailboxRightRequest(sessionFixture, accountIdFixture, request))
+          .thenAnswer((_) async => true);
+
+        when(mailboxDataSource.getChanges(sessionFixture, accountIdFixture, StateFixtures.currentMailboxState))
+          .thenAnswer((_) async => MailboxChangeResponse(
+              // dont really know what to put in there
+              updated: [mailboxFixture],
+              updatedProperties: Properties({'rights'})
+        ));
+
+        final response = await mailboxDataSource.getChanges(
+          sessionFixture,
+          accountIdFixture,
+          StateFixtures.currentMailboxState,
+        );
```
But this test is meaningless. You mock it, then test the mock directly :))

florentos17 replied:
yes, it is what i said here: https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1856180497

but i am about to drop [the commit with these changes](https://github.com/linagora/tmail-flutter/pull/3241/commits/767da25dd7705c096242f6718d2997024eecada0). cf the comments linked above

florentos17 replied:
done in `test/features/mailbox/data/update_rights_for_subaddressing_test.dart` !

florentos17 commented:
>     * [ ]  how to test sub-addressing (tmail.linagora can not work when sending email to subaddress email)

you need to connect the frontend to a server that has the latest changes, including  [this PR](https://github.com/apache/james-project/pull/2473), and [this PR](https://github.com/apache/james-project/pull/2491). when running my own instance of the server, with the image `linagora/tmail-backend-distributed`, it works.

note that subaddressing features only show if the server advertizes that it supports it. but currently that advertizement is not merged on the server side (cf [this PR](https://github.com/apache/james-project/pull/2521)) so you'll have to manually change the default value to true in `lib/features/mailbox/presentation/mixin/mailbox_widget_mixin.dart` line `290` before running the front.

>     * [ ]  generating MailAddress for sub-address (encode, decode) + testing all cases ([ISSUE-3189 front end support for subaddressing #3241 (comment)](https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1861499772))

i will post a video testing all cases

>     * [ ]  unusefull test case [ISSUE-3189 front end support for subaddressing #3241 (comment)](https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1862929054)

done

>     * [ ]  analyze

i feel like i have a test failing, but i can't find which one:
```
Run ./scripts/test.sh
+ [[ default == \d\e\f\a\u\l\t ]]
+ flutter test -r json
Error: Process completed with exit code 1.
```

hoangdat commented:
- [ ] error in test
<img width="697" alt="Screenshot 2024-12-02 at 09 10 07" src="https://github.com/user-attachments/assets/7beba69b-824a-477a-abd8-06be3e12cdb4">

hoangdat commented on lib/features/mailbox/presentation/mixin/mailbox_widget_mixin.dart:
```diff
@@ -248,6 +273,23 @@ mixin MailboxWidgetMixin {
       .toList();
   }
 
+  bool isSubaddressingSupported(Session? session, AccountId? accountId) {
+    if (session == null || accountId == null) {
+      return false;
+    }
+    if (!CapabilityIdentifier.jmapTeamMailboxes.isSupported(session, accountId)) {
+      return false;
+    }
+
+    final subaddressingSupported = ((session
+        .props[0] as Map<CapabilityIdentifier, CapabilityProperties>?)
+        ?[CapabilityIdentifier.jmapTeamMailboxes]
+        ?.props[0] as Map<String, dynamic>?)
+        ?['subaddressingSupported'];
+
+    return subaddressingSupported ?? false;
+  }
+
```
Should write test for this. It is not run well with 
```
{
    "capabilities": {
        "urn:apache:james:params:jmap:delegation": {},
        "com:linagora:params:jmap:pgp": {},
        "urn:ietf:params:jmap:websocket": {
            "supportsPush": true,
            "url": "wss://jmap.linagora.com/oidc/jmap/ws"
        },
        "com:linagora:params:jmap:settings": {},
        "com:linagora:params:jmap:ws:ticket": {
            "generationEndpoint": "https://jmap.linagora.com/oidc/jmap/ws/ticket",
            "revocationEndpoint": "https://jmap.linagora.com/oidc/jmap/ws/ticket"
        },
        "urn:apache:james:params:jmap:mail:shares": {},
        "urn:ietf:params:jmap:vacationresponse": {},
        "urn:ietf:params:jmap:mail": {
            "maxMailboxesPerEmail": 10000000,
            "maxMailboxDepth": null,
            "maxSizeMailboxName": 200,
            "maxSizeAttachmentsPerEmail": 20971520,
            "emailQuerySortOptions": [
                "receivedAt",
                "sentAt",
                "size",
                "from",
                "to",
                "subject"
            ],
            "mayCreateTopLevelMailbox": true
        },
        "com:linagora:params:jmap:labels": {},
        "urn:ietf:params:jmap:mdn": {},
        "urn:apache:james:params:jmap:mail:quota": {},
        "urn:ietf:params:jmap:quota": {},
        "urn:apache:james:params:jmap:mail:identity:sortorder": {},
        "urn:ietf:params:jmap:submission": {
            "maxDelayedSend": 0,
            "submissionExtensions": {}
        },
        "com:linagora:params:calendar:event": {
            "replySupportedLanguage": [
                "en",
                "fr"
            ]
        },
        "urn:ietf:params:jmap:core": {
            "maxSizeUpload": 20971520,
            "maxConcurrentUpload": 4,
            "maxSizeRequest": 10000000,
            "maxConcurrentRequests": 4,
            "maxCallsInRequest": 16,
            "maxObjectsInGet": 500,
            "maxObjectsInSet": 500,
            "collationAlgorithms": [
                "i;unicode-casemap"
            ]
        },
        "com:linagora:params:jmap:echo": {},
        "com:linagora:params:jmap:team:mailboxes": {},
        "com:linagora:params:jmap:forward": {},
        "com:linagora:params:jmap:contact:support": {
            "supportMailAddress": null
        },
        "com:linagora:params:jmap:contact:autocomplete": {
            "minInputLength": 2
        },
        "com:linagora:params:jmap:filter": {},
        "com:linagora:params:jmap:firebase:push": {
            "apiKey": null,
            "appId": null,
            "messagingSenderId": null,
            "projectId": null,
            "databaseUrl": null,
            "storageBucket": null,
            "authDomain": null,
            "vapidPublicKey": null
        },
        "com:linagora:params:jmap:public:assets": {
            "publicAssetTotalSize": 20971520
        },
        "com:linagora:params:jmap:messages:vault": {
            "maxEmailRecoveryPerRequest": "5",
            "restorationHorizon": "15 days"
        }
    },
    "accounts": {
        "xxx": {
            "name": "dphamhoang@linagora.com",
            "isPersonal": true,
            "isReadOnly": false,
            "accountCapabilities": {
                "urn:apache:james:params:jmap:delegation": {},
                "com:linagora:params:jmap:pgp": {},
                "urn:ietf:params:jmap:websocket": {
                    "supportsPush": true,
                    "url": "wss://jmap.linagora.com/oidc/jmap/ws"
                },
                "com:linagora:params:jmap:settings": {},
                "com:linagora:params:jmap:ws:ticket": {
                    "generationEndpoint": "https://jmap.linagora.com/oidc/jmap/ws/ticket",
                    "revocationEndpoint": "https://jmap.linagora.com/oidc/jmap/ws/ticket"
                },
                "urn:apache:james:params:jmap:mail:shares": {},
                "urn:ietf:params:jmap:vacationresponse": {},
                "urn:ietf:params:jmap:mail": {
                    "maxMailboxesPerEmail": 10000000,
                    "maxMailboxDepth": null,
                    "maxSizeMailboxName": 200,
                    "maxSizeAttachmentsPerEmail": 20971520,
                    "emailQuerySortOptions": [
                        "receivedAt",
                        "sentAt",
                        "size",
                        "from",
                        "to",
                        "subject"
                    ],
                    "mayCreateTopLevelMailbox": true
                },
                "com:linagora:params:jmap:labels": {},
                "urn:ietf:params:jmap:mdn": {},
                "urn:apache:james:params:jmap:mail:quota": {},
                "urn:ietf:params:jmap:quota": {},
                "urn:apache:james:params:jmap:mail:identity:sortorder": {},
                "urn:ietf:params:jmap:submission": {
                    "maxDelayedSend": 0,
                    "submissionExtensions": {}
                },
                "com:linagora:params:calendar:event": {
                    "replySupportedLanguage": [
                        "en",
                        "fr"
                    ]
                },
                "urn:ietf:params:jmap:core": {
                    "maxSizeUpload": 20971520,
                    "maxConcurrentUpload": 4,
                    "maxSizeRequest": 10000000,
                    "maxConcurrentRequests": 4,
                    "maxCallsInRequest": 16,
                    "maxObjectsInGet": 500,
                    "maxObjectsInSet": 500,
                    "collationAlgorithms": [
                        "i;unicode-casemap"
                    ]
                },
                "com:linagora:params:jmap:echo": {},
                "com:linagora:params:jmap:team:mailboxes": {},
                "com:linagora:params:jmap:forward": {},
                "com:linagora:params:jmap:contact:support": {
                    "supportMailAddress": null
                },
                "com:linagora:params:jmap:contact:autocomplete": {
                    "minInputLength": 2
                },
                "com:linagora:params:jmap:filter": {},
                "com:linagora:params:jmap:firebase:push": {
                    "apiKey": null,
                    "appId": null,
                    "messagingSenderId": null,
                    "projectId": null,
                    "databaseUrl": null,
                    "storageBucket": null,
                    "authDomain": null,
                    "vapidPublicKey": null
                },
                "com:linagora:params:jmap:public:assets": {
                    "publicAssetTotalSize": 20971520
                },
                "com:linagora:params:jmap:messages:vault": {
                    "maxEmailRecoveryPerRequest": "5",
                    "restorationHorizon": "15 days"
                }
            }
        }
    },
    "primaryAccounts": {
        "urn:apache:james:params:jmap:delegation": "xxx",
        "com:linagora:params:jmap:pgp": "xxx",
        "urn:ietf:params:jmap:websocket": "xxx",
        "com:linagora:params:jmap:settings": "xxx",
        "com:linagora:params:jmap:ws:ticket": "xxx",
        "urn:apache:james:params:jmap:mail:shares": "xxx",
        "urn:ietf:params:jmap:vacationresponse": "xxx",
        "urn:ietf:params:jmap:mail": "xxx",
        "com:linagora:params:jmap:labels": "xxx",
        "urn:ietf:params:jmap:mdn": "xxx",
        "urn:apache:james:params:jmap:mail:quota": "xxx",
        "urn:ietf:params:jmap:quota": "xxx",
        "urn:apache:james:params:jmap:mail:identity:sortorder": "xxx",
        "urn:ietf:params:jmap:submission": "xxx",
        "com:linagora:params:calendar:event": "xxx",
        "urn:ietf:params:jmap:core": "xxx",
        "com:linagora:params:jmap:echo": "xxx",
        "com:linagora:params:jmap:team:mailboxes": "xxx",
        "com:linagora:params:jmap:forward": "xxxx",
        "com:linagora:params:jmap:contact:support": "xxx",
        "com:linagora:params:jmap:contact:autocomplete": "xxx",
        "com:linagora:params:jmap:filter": "xxxx",
        "com:linagora:params:jmap:firebase:push": "xxxx",
        "com:linagora:params:jmap:public:assets": "xxx",
        "com:linagora:params:jmap:messages:vault": "xxxx"
    },
    "username": "dphamhoang@linagora.com",
    "apiUrl": "https://jmap.linagora.com/oidc/jmap",
    "downloadUrl": "https://jmap.linagora.com/oidc/download/{accountId}/{blobId}?type={type}&name={name}",
    "uploadUrl": "https://jmap.linagora.com/oidc/upload/{accountId}",
    "eventSourceUrl": "https://jmap.linagora.com/oidc/eventSource?types={types}&closeAfter={closeafter}&ping={ping}",
    "state": "2c9f1b12-b35a-43e6-9af2-0106fb53a943"
}
```

hoangdat replied:
please use `session_extension::getCapabilityProperties` to make sure we get properties not only in `capabilities`, but also in `accounts` instance.

florentos17 replied:
you `session` attribute does not advertize that subaddressing is supported so in your case `isSubaddressingSupported` should return `false`

would you prefer
```
final subaddressingSupported = (session.getCapabilityProperties(accountId, CapabilityIdentifier.jmapTeamMailboxes) as Map<String, dynamic>)['subaddressingSupported'];
 return subaddressingSupported2 ?? false;
```

hoangdat replied:
Sure this session not support subaddressing, but still see the menu action for it. Please check again. Write tests for this too

hoangdat commented:
- [ ] Please let me know how I can verify your completion and corrections. Maybe using dev (mai.lin-saas.dev) would be okay.

dab246 commented on lib/features/caching/caching_manager.dart:
```diff
@@ -118,6 +118,15 @@ class CachingManager {
     ], eagerError: true);
   }
 
+  Future<void> clearEmailCacheAndAllStateCacheAndMailboxCache() {
+    return Future.wait([
+      _stateCacheClient.clearAllData(),
+      _emailCacheClient.clearAllData(),
+      _mailboxCacheClient.clearAllData(),
+      if (PlatformInfo.isMobile) clearAllFileInStorage(),
+    ], eagerError: true);
+  }
```
IMO, we don't need remove data cache

hoangdat replied:
he upgrade model for Mailbox, so I request to migrate data. Why we dont need?

dab246 replied:
So we just only remove mailbox cache.

florentos17 commented:
> > * [ ]   generating MailAddress for sub-address (encode, decode) + testing all cases ([ISSUE-3189 front end support for subaddressing #3241 (comment)](https://github.com/linagora/tmail-flutter/pull/3241#discussion_r1861499772))
> 
> i will post a video testing all cases

here is a little demo video testing several edge cases:


https://github.com/user-attachments/assets/fe179142-1d6a-4a8d-a6d7-19bb1b5910f8

dab246 commented on lib/features/caching/caching_manager.dart:
```diff
@@ -118,6 +118,13 @@ class CachingManager {
+  Future<void> clearMailboxCache() {
+    return Future.wait([
+      _mailboxCacheClient.clearAllData(),
+      if (PlatformInfo.isMobile) clearAllFileInStorage(),
```
Please remove it, we only need remove mailbox cache and state of mailbox

florentos17 replied:
so just this ?

```
Future<void> clearMailboxCache() {
    return Future.wait([
      _mailboxCacheClient.clearAllData(),
      _stateCacheClient.clearAllData()
    ], eagerError: true);
  }
  ```

hoangdat replied:
stateCache store the data via `tuppleKey`. So we only to clear state of mailbox. Please try with
```
_stateCacheClient.deleteItem(StateType.mailbox.getTupleKeyStored(accountId, session.username))
```

hoangdat replied:
okie so please 
```
Future<void> clearMailboxCache() {
  return Future.wait([
    _mailboxCacheClient.clearAllData(),
  ], eagerError: true);
}
```

florentos17 replied:
done !

dab246 commented on lib/features/mailbox/presentation/mixin/mailbox_widget_mixin.dart:
```diff
@@ -248,6 +270,20 @@ mixin MailboxWidgetMixin {
+
+    return (session.getCapabilityProperties(accountId, CapabilityIdentifier.jmapTeamMailboxes)
+        ?.props[0] as Map<String, dynamic>?)
+        ?['subaddressingSupported']
```
Add it to `MailboxConstants`

florentos17 replied:
done

dab246 commented on test/features/mailbox/presentation/mailbox_widget_mixin_test.dart:
```diff
@@ -0,0 +1,69 @@
+void main() {
+  group('MailboxWidgetMixin::isSubaddressingSupported::test', () {
+
+    test(
+        'should return true '
+            'when the server advertizes true',
+            () {
```
Pay attention to code alignment

dab246 commented on test/features/mailbox/presentation/mailbox_widget_mixin_test.dart:
```diff
@@ -0,0 +1,69 @@
+        });
+
+    test(
+        'should return false '
+            'when the server advertizes nothing',
+            () {
```
idem

dab246 reviewed (approved):
LGTM
